### ISSUE TITLE
Add support for UTC cron strings

### ISSUE THREAD

**Issue by dprentis** [harrisiirak/cron-parser#44]
Would be useful to be able to interpret cron strings as being in UTC rather than local time.

**Comment by harrisiirak** [comment]
Hey Daniel!

What do you exactly mean? You can provide `startDate` and `endDate` as you wish in the parser options.

BR

**Comment by dprentis** [comment]
Well, for example if I have this cron strong: 0 0 10 * * *, then I need it to be interpreted as 10:00 am UTC. Currently it gets interpreted according to the computers local time, so if it's currently 8:00 am local time, then next() will return a date with 10:00 am local time (which could be anything compared to UTC, depending on where I'm located).

Use case is that I want to use persisted cron strings in a distributed system, so I need UTC as a single reference point.

later.js, for example, provides a global method to specify whether to use UTC or not http://bunkat.github.io/later/occurrences.html

To be honest, I don't think a global method is a good idea, but I'm currently using later.js because I haven't found any alternative.

Would be great if I could just do something like:

    parser.parseExpression('0 0 10 * * *', {utc: true});

Otherwise a great lib that does one thing - thanks! :)

**Comment by harrisiirak** [comment]
Yes and no, I could add timezone support (not only UTC!), but this currently achievable by providing your converted date (in UTC) as `startDate` option value.

I must think about adding timezone support, because ideally I want to keep this library as small as possible and keep it dependency free. So I still would recommend to use external library to do date conversion from one timezone to another. For example I think moment.js is quite nice one for that.

**Comment by dprentis** [comment]
Ok, not sure I follow how startDate affects the way the cron string is interpreted. Doesn't it just denote start of the interval in which iterations can happen? Maybe I can try this later, but I was under the impression that next always returns a local time according to the cron string.

Converting the date from one timezone to another is not the issue - yes, I could add x hours to the returned value of next() to compensate the timezone. But then I would have to differentiate "hour dependent" strings (eg. "0 0 10 * * *" -> need 10:00am UTC), from "hour independent" strings (eg. "0 */5 * * * *" -> need some time within next 5 minutes) - basically I'd have to do some parsing, which would pretty well defeat the purpose of using a cron parsing lib in the first place.

AFAIK, adding support for UTC would not add any dependencies - you would probably just need to use the UTC Date methods. Yes, adding general time zones would be more complex, but that's probably overkill.

Maybe I could try a pull request when I get the time (no pun intended).

**Comment by harrisiirak** [comment]
Yes, you're correct, by the default it always returns local time, expect when you provide `startDate` which denotes the beginning of the iteration.

If you meant that ` utc` option flag just sets local time automatically to UTC then that still can achieved by setting `startDate` yourself. This option can be added, but that doesn't change much, only making things a little bit cleaner. The logic you described is anyway quite application specific: I think for you first use case utc = true and second one utc = false?

PR's are always welcome and probably I'm going to understand better what are meaning here.

**Comment by dprentis** [comment]
Ok, maybe an example would help - can you give me an example in code of how to set startDate to interpret the cron string as UTC?

eg. I have "0 0 10 * * *" and "0 */5 * * *" In both cases I want to get the next iteration (from now) as a date.

In the first case the timezone is important, because a specific hour is specified - I want "0 0 10 * * *" to be interpreted as UTC (in my particular timezone, it should compute to 12:00 midday local time).

In the second case timezone is irrelevant, because the schedule happens every 5 minutes, but I need the next iteration - so any timezone adjustment would mess things up.

I need a single piece of code to handle both cases, so I don't have to differentiate manually by parsing the strings - what do I do?

**Comment by harrisiirak** [comment]
I think I got what you meant. Anyhow, it was also good time to clean up Date class pollution. From now on `next()` method returns extended Date instance (can't extend Date class directly so doing this in a little bit other way) what has `toUTC()` method. You have no need specify any extra configuration option.

See README for updated usage.

PR: https://github.com/harrisiirak/cron-parser/pull/45
Branch: https://github.com/harrisiirak/cron-parser/tree/feature/date-class-pollution-cleanup

Let me know if you have ideas for improvements.

**Comment by dprentis** [comment]
ok, thanks for your time! I haven't had time to check this through properly or test it.

IMHO, I would probably stick to returning a standard Date if possible, implement any internally used Date methods as functions that take Date as a first parameter, to prevent polluting the Date prototype.

I am not sure how the toUTC method solves the problem - looks like it returns a Date-like (CronDate) object with time shifted by timezone. AFAIK, that's something I can do quite easily with a standard Date object or using moment for example. Issue is: how can I know when I should shift and when I shouldn't shift, for any given cron string?

I'm probably missing something. Wish I could get you a PR - will try to find some time this weekend - wife permitting..

**Comment by harrisiirak** [comment]
> IMHO, I would probably stick to returning a standard Date if possible, implement any internally used Date methods as functions that take Date as a first parameter, to prevent polluting the Date prototype.

Actually you're getting back  `Date ` instance but with some additional methods.

> I am not sure how the toUTC method solves the problem - looks like it returns a Date-like (CronDate) object with time shifted by timezone.

Isn't that what you want in your first case?

>  Issue is: how can I know when I should shift and when I shouldn't shift, for any given cron string?

As you described your uses cases... isn't that application itself that should decided what logic in what context it should use? Basically if you want use it locally use it as is. And if you want universal time version you call  `toUTC() ` which returns shifted  `Date` instance. Or you can always  use `Date ` class standard  `toISOString() ` or  `toUTCString() ` to get universal time string representation.

**Comment by dprentis** [comment]
So finally got that pull request to you: https://github.com/harrisiirak/cron-parser/pull/46

Basically, setting utc to true in options will cause the schedule calculation to use the \*UTC\* Date object methods instead of the local time ones.

All tests pass. I added a new unit test, which reflects the basic example here - I expect (hope) other cron strings will 'just work', but let me know if you see anything obviously wrong.

**Comment by dprentis** [comment]
>As you described your uses cases... isn't that application itself that should decided what logic in what context it should use? Basically if you want use it locally use it as is. And if you want universal time version you call toUTC() which returns shifted Date instance. Or you can always use Date class standard toISOString() or toUTCString() to get universal time string representation.

Hopefully my pull request will make things more apparent - try running my unit test with the code base in master. You should see that if you don't shift, the first check will fail. If you do shift in both cases, the second check will fail.

Basically the app wants to deal with cron strings that reflect the time in UTC and it doesn't want to have to decide if it should shift the result or not, because that would require parsing the string - some results must be shifted while others must not be shifted.

```
test('parse expression as UTC', function(t) {
  try {
    var options = {
      utc: true
    };

    var interval = CronExpression.parse('0 0 10 * * *', options);

    var date = interval.next();
    t.equal(date.getUTCHours(), 10, 'Correct UTC hour value'); // would have had to shift

    interval = CronExpression.parse('0 */5 * * * *', options);

    var date = interval.next(), now = new Date();
    now.setMinutes(now.getMinutes() + 5 - (now.getMinutes() % 5));

    // no shift here, regardless of whether cron string is UTC or not
    t.equal(date.getHours(), now.getHours(), 'Correct local time for 5 minute interval');

  } catch (err) {
    t.ifError(err, 'UTC parse error');
  }

  t.end();
});
```

**Comment by harrisiirak** [comment]
Yep, that makes a little bit more sense. Thanks, looks good! Going to look into that more deeply on tomorrow's evening or Monday.

Still going to merge #45 too, because it also addresses some other issues. Let's see in which order I'm going to merge them (there shouldn't be any conflicts thou).

**Comment by harrisiirak** [comment]
Merged #44 and #45
